apps/s_server.c

```diff
@@ -3330,6 +3348,43 @@ static int www_body(int s, int stype, int prot, unsigned char *context)
                              "HTTP/1.0 200 ok\r\nContent-type: text/plain\r\n\r\n");
             }
             /* send the file */
+#ifndef OPENSSL_NO_KTLS
+            if (use_sendfile) {
+                FILE *fp = NULL;
+                int fd;
+                struct stat st;
+                off_t offset = 0;
+                size_t sz;
+
+                BIO_get_fp(file, &fp);
+                fd = fileno(fp);
+                if (fstat(fd, &st) < 0) {
+                    BIO_printf(io, "Error stat '%s'\r\n", p);
+                    goto write_error;
+                }
+
+                sz = st.st_size;
+                if (((int)BIO_flush(io)) < 0)
+                    goto write_error;
+
+                for (;;) {
+                    i = SSL_sendfile(con, fd, offset, sz, 0);
+                    if (i < 0) {
+                        BIO_printf(io, "Error SSL_sendfile '%s'\r\n", p);
+                        ERR_print_errors(io);
+                        break;
+                    } else {
+                        offset += i;
+                        sz -= i;
+                    }
+
+                    if (sz <= 0) {
+                        BIO_printf(bio_err, "KTLS SENDFILE '%s' OK\n", p);
+                        break;
+                    }
+                }
+            } else
```

Comment:
This `} else` is not correct, lacks of braces

Reply:
There is no problem here, but the code behind else is not indented, which is also to modify the code as little as possible

Reply:
But this is very subtle, since if it's changed here, the a final brace is also to be added... not sure if this is worth a fix...

Reply:
There is only one 'for' statement after the 'else', and it is not necessary to add a brace in syntax. If the code style of openssl needs it, I will add a brace.

Reply:
Yes, typically OpenSSL coding style requires if the `if` block has a pair of braces, then the `else` should have one as well. But this situation is different due to the `#end` stuffs involved.

Reply:
consider duplicating code (which will only exist in one #ifdef branch).  That brings its own risk, but I think the consistent brace/indent is more important for developer readability.

Reply:
So the suggestion is something like this?

~~~
#ifndef OPENSSL_NO_KTLS
    if (use_sendfile) {
        ...
    } else {
        for (;;) {
        ...
    }
#else 
    for (;;) {
    ...
#endif 
~~~

Reply:
yeah.  and make the body of the for loop a macro if you want.

Reply:
This will introduce redundant code. Are you sure you want to do this? @mattcaswell , What's your opinion？

Reply:
Yes it will introduce redundant code.  That can be mitigated by using a #define for the body of the for loop.  The trade-off is that the code is consistent and understandable.

Reply:
I don't agree with @richsalz here. If it was a line or two I could live with the duplication, but this is more than that. I would prefer:

````
#ifndef OPENSSL_NO_KTLS
    if (use_sendfile) {
        ...
    } else
#endif
    {
        for (;;) {
        ...
    }
````

Reply:
Matt's on the project, I'm not, so listen to him not me :)

Reply:
> 
> 
> Matt's on the project, I'm not, so listen to him not me :)

Also thanks also for your suggestions.